### ISSUE TITLE
Cannot add a bridged Accessory with the same UUID as another bridged Accessory: undefined

### ISSUE THREAD

**Issue by macinspak** [homebridge/HAP-NodeJS#872]
**Describe the bug**
Difficult to reproduce, I have it currently. I noticed it first in node-red-contrib-homekit-bridged, but after debugging for a while, it seems that although NRCHKB is only calling AccessoryUtils.getOrCreate once, the loop **inside HAP-NodeJS** to add it internally to the accessories in the bridge is looping twice and seems to have an index of "NaN" already in the fresh bridge with an empty array of accessories, which is causing it to "match" a non-existent UUID in the bridge and complain about an undefined UUID (when the UUID is defined in the accessory).

This seemed to "appear" in an installation I had been running successfully for a long time, then (all of a sudden, no changes, yeah- I know) and I was getting "no response" on all devices in homekit, so I removed the bridge from home.app but now cannot re-add the bridge, it doesn't appear in the home.app at all.

a lot of debug leading to this issue being raised is documented in [NRCHKB Issue 359](https://github.com/NRCHKB/node-red-contrib-homekit-bridged/issues/359)

**To Reproduce**
Seems to be only on my server, someone else running the same flow (as per [test flow](https://github.com/NRCHKB/node-red-contrib-homekit-bridged/issues/359#issuecomment-758411019)) seems to not get the issue.

**Desktop (please complete the following information):**
 - OS: Rpi
 - Node version: v14.15.4
 - NPM version: 6.14.10
 - NodeRed version: v1.2.7
 - NRCHKB Plugin Version: 1.2.0
 - NRCHKB uses HAP-NodeJS Version: 0.7.8

**Additional context**
running debug

```
DEBUG=NRCHKB*,HAP-NodeJS*,Accessory,HAPServer,EventedHTTPServer node-red
```

I get a lot of these first in the debug output:

```
  NRCHKB:HAPServiceNode Starting Parent Service Pool 1 +6ms
  NRCHKB:AccessoryUtils Looking for accessory with service subtype '172f1edd-389b-4c45-96d4-47e37a130118'... +0ms
  NRCHKB:AccessoryUtils ... didn't find it. Adding new accessory with name 'Pool 1' and UUID '1316cdc7-e142-4a55-9a93-fea8fce35aa5' +0ms
  NRCHKB:HAPServiceNode Starting Parent Service Pool 2 +6ms
  NRCHKB:AccessoryUtils Looking for accessory with service subtype 'fd02bca6-d5c5-4777-9d45-73cea71e4b6d'... +0ms
  NRCHKB:AccessoryUtils ... didn't find it. Adding new accessory with name 'Pool 2' and UUID '4a78594b-60fa-470d-a192-cb14e75910e6' +1ms
```

Then a lot of these:
```
(node:30289) UnhandledPromiseRejectionWarning: Error: Cannot add a bridged Accessory with the same UUID as another bridged Accessory: undefined
    at Bridge.Accessory._this.addBridgedAccessory (/home/pi/.node-red/node_modules/hap-nodejs/dist/lib/Accessory.js:286:27)
    at Bridge.Accessory._this.addBridgedAccessories (/home/pi/.node-red/node_modules/hap-nodejs/dist/lib/Accessory.js:311:23)
    at Object.getOrCreate (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/utils/AccessoryUtils.js:80:20)
    at configure (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:118:40)
    at preInit.init (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:38:13)
    at /home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:28:18
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:30289) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 21)

```

Note the UUID is shown as **undefined** - this is not correct, it is most certainly defined.

For more detail on the existing debugging leading to this conclusion, please refer to [NRCHKB Issue 359](https://github.com/NRCHKB/node-red-contrib-homekit-bridged/issues/359)

I think some config is "stuck" somewhere that I cannot seem to find, that pre-fills an "empty" new bridge accessory array with a "NaN" index. I think it's worth debugging this to first resolve it and then work out how it could have happened, to ensure it doesn't happen for anyone else.

**Comment by macinspak** [comment]
add debug documentation link to another issue

**Comment by macinspak** [comment]
Any thoughts?

**Comment by Shaquu** [comment]
I still see it is difficult to investigate.
@macinspak can you start clean? New OS, fresh installation of everything.

**Comment by otaku** [comment]
I ran into this as well.  Imported my flows into a new instance and everything ran without encountering any errors.  Ended up erasing my node-red folder and importing my flows again.

**Comment by macinspak** [comment]
> I still see it is difficult to investigate.
> @macinspak can you start clean? New OS, fresh installation of everything.

@Shaquu I can, but that just means someone else will get the problem right? In fact it seems this will keep happening unless we can get an environment where it is reproducable. I have one now that is 100% reproducable and happy to work with anyone to work out the issue so we can fix it so it never happens to anyone again.

As @otaku mentioned, he's seen it too. But where is it coming from and why does a complete new install and reimport flows work? I cannot find any remnance of the nodes in any persisted files, unless it's something in the flow_creds.json or a hidden .settings or something

**Comment by macinspak** [comment]
From what I've seen, this happens on various occasions, so we should try and fix it instead of just "reinstall"

**Comment by macinspak** [comment]
Without me reverse engineering hap-nodejs from the start, what is expected to be in the accessory array for a newly created bridge on first call to add accessory?

I mean, can someone print out the index array from a working installation? Does it have "NaN" as an index in the empty array?

Secondly, as I mentioned in the other thread, it actually goes through adding each one, but says it cannot.
ie:

add accessory1
in the addAccessory routine this.bridgedAccessories = [] and the indexes to that array are ['Nan']
add accessory2
in the addAccessory routine this.bridgedAccessories = ['accessory1'] and the indexes to that array are [0, 'Nan']
add accessory3
in the addAccessory routine this.bridgedAccessories = ['accessory1', 'accessory2'] and the indexes to that array are [0, 1, 'Nan']

So for some reason, the new bridge.bridgedAccessories array always has one extra index, being 'NaN'. Is this an issue? or is it as designed? Or does no one else see it? If so, then why do I have it?

As mentioned in the other issue, if I tweak the check code to ignore the "NaN" index, it gets past that problem, but then throws a new issue:

```
(node:3775) UnhandledPromiseRejectionWarning: TypeError: accessory.on is not a function
    at Bridge.Accessory._this.addBridgedAccessory (/home/pi/.node-red/node_modules/hap-nodejs/dist/lib/Accessory.js:297:23)
    at Bridge.Accessory._this.addBridgedAccessories (/home/pi/.node-red/node_modules/hap-nodejs/dist/lib/Accessory.js:315:23)
    at Object.getOrCreate (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/utils/AccessoryUtils.js:80:20)
    at configure (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:118:40)
    at preInit.init (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:38:13)
    at /home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:28:18
    at emitUnhandledRejectionWarning (internal/process/promises.js:168:15)
    at processPromiseRejections (internal/process/promises.js:247:11)
    at processTicksAndRejections (internal/process/task_queues.js:94:32)
(node:3775) TypeError: accessory.on is not a function
    at Bridge.Accessory._this.addBridgedAccessory (/home/pi/.node-red/node_modules/hap-nodejs/dist/lib/Accessory.js:297:23)
    at Bridge.Accessory._this.addBridgedAccessories (/home/pi/.node-red/node_modules/hap-nodejs/dist/lib/Accessory.js:315:23)
    at Object.getOrCreate (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/utils/AccessoryUtils.js:80:20)
    at configure (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:118:40)
    at preInit.init (/home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:38:13)
    at /home/pi/.node-red/node_modules/node-red-contrib-homekit-bridged/build/lib/HAPServiceNode.js:28:18
```

**Comment by Shaquu** [comment]
@macinspak imho reproducible would be if you would be able to give me steps that I can follow to see the same error on my device :)

I am not forcing you to stop investigating though.
I just never seen it.

Regarding what @otaku said.
Recently we had a situation where one of the other nodes (then nrchkb) in the flow was causing whole node-red to not work correctly so maybe indeed cherry picking from other flows might be a good idea to find an issue?

**Comment by macinspak** [comment]
Right, I've decided to debug myself.

So made a little progress. I checked the array being passed by NRCHKB (yes, that's what I'm using) into bridge.addAccessories and it APPEARS to be a single element array as expected.

So I changed the 2 loops in both addAccessories and addAccessory (in hap-nodejs) from:
```javascript
for (var index in accessories) {
    var accessory = accessories[index];
    _this.addBridgedAccessory(accessory, true);
}
```
to 
```javascript
accessories.forEach(function(accessory) {
    _this.addBridgedAccessory(accessory, true);
});
```

and 

```javascript
for (var index in accessories) {
    var accessory = accessories[index];
    _this.addBridgedAccessory(accessory, true);
}
```
to 
```javascript
accessories.forEach(function(accessory) {
    _this.addBridgedAccessory(accessory, true);
});
```

and IT STARTED WORKING (well mostly...)

It now adds all accessories and reports all is fine :
```
  NRCHKB:HAPBridgeNode Setting name to Homered Bridge +0ms
  NRCHKB:HAPBridgeNode Creating Bridge with name 'Homered Bridge' and UUID 'f01002cf-958b-4a49-8c39-bca1d69ad92b' +3ms
9 Feb 11:49:25 - [info] [AnamicoLightFxChristmas:2c7415d7.8b8eaa] festive = false
9 Feb 11:49:25 - [info] Started flows
  NRCHKB:HAPServiceNode Starting Parent Service Alarm +0ms
  NRCHKB:HAPServiceNode getOrCreate Accessory: {"name":"Alarm","UUID":"e54367ad-f42b-44ca-8869-de169a41dba9","manufacturer":"NRCHKB","serialNo":"Default Serial Number","model":"1.2.0","firmwareRev":"1.2.0","hardwareRev":"1.2.0","softwareRev":"1.2.0"} +12ms
  NRCHKB:AccessoryUtils Looking for accessory with service subtype 'cf2eddfe-0672-4b7a-9eb8-a9dc7d694232'... +0ms
  NRCHKB:AccessoryUtils ... didn't find it. Adding new accessory with name 'Alarm' and UUID 'e54367ad-f42b-44ca-8869-de169a41dba9' +1ms
  HAP-NodeJS:Accessory [Alarm] Identification request +0ms
  HAP-NodeJS:Accessory [Alarm] Identification request ignored; no listeners to 'identify' event +1ms
  NRCHKB:AccessoryUtils Bridge now has 1 accessories. +34ms
  NRCHKB:HAPServiceNode getOrCreate Service: {"name":"Alarm","UUID":"cf2eddfe-0672-4b7a-9eb8-a9dc7d694232","serviceName":"SecuritySystem","config":{"id":"75c51002.20d88","type":"homekit-service","z":"ffeccbc0.842cc8","g":"3e2e8860.d3bec8","isParent":true,"bridge":"8cb8961e.cbe378","parentService":"","name":"Alarm","serviceName":"SecuritySystem","topic":"","filter":false,"manufacturer":"NRCHKB","model":"1.2.0","serialNo":"Default Serial Number","firmwareRev":"1.2.0","hardwareRev":"1.2.0","softwareRev":"1.2.0","cameraConfigVideoProcessor":"ffmpeg","cameraConfigSource":"","cameraConfigStillImageSource":"","cameraConfigMaxStreams":2,"cameraConfigMaxWidth":1280,"cameraConfigMaxHeight":720,"cameraConfigMaxFPS":10,"cameraConfigMaxBitrate":300,"cameraConfigVideoCodec":"libx264","cameraConfigAudioCodec":"libfdk_aac","cameraConfigAudio":false,"cameraConfigPacketSize":1316,"cameraConfigVerticalFlip":false,"cameraConfigHorizontalFlip":false,"cameraConfigMapVideo":"0:0","cameraConfigMapAudio":"0:1","cameraConfigVideoFilter":"scale=1280:720","cameraConfigAdditionalCommandLine":"-tune zerolatency","cameraConfigDebug":false,"cameraConfigSnapshotOutput":"disabled","cameraConfigInterfaceName":"","characteristicProperties":"{}","waitForSetupMsg":false,"outputs":2,"x":580,"y":220,"wires":[["e8aee56b.c75b98"],[]]}} +38ms
  NRCHKB:ServiceUtils Looking for service with UUID 'cf2eddfe-0672-4b7a-9eb8-a9dc7d694232'... +0ms
  NRCHKB:ServiceUtils ... didn't find it. Adding new SecuritySystem service. +1ms
  NRCHKB:HAPServiceNode Starting Parent Service Ensuite Glass +8ms
  NRCHKB:HAPServiceNode getOrCreate Accessory: {"name":"Ensuite Glass","UUID":"55a4856d-b41c-4b98-a0dc-2179d307a5ff","manufacturer":"NRCHKB","serialNo":"Default Serial Number","model":"0.130.29","firmwareRev":"0.130.29","hardwareRev":"0.130.29","softwareRev":"0.130.29"} +3ms
  NRCHKB:AccessoryUtils Looking for accessory with service subtype '1803449b-0cc0-4cd7-8139-6e853b07d5e3'... +0ms
  NRCHKB:AccessoryUtils ... didn't find it. Adding new accessory with name 'Ensuite Glass' and UUID '55a4856d-b41c-4b98-a0dc-2179d307a5ff' +0ms
  HAP-NodeJS:Accessory [Ensuite Glass] Identification request +32ms
  HAP-NodeJS:Accessory [Ensuite Glass] Identification request ignored; no listeners to 'identify' event +1ms
  NRCHKB:AccessoryUtils Bridge now has 2 accessories. +18ms
  NRCHKB:HAPServiceNode getOrCreate Service: {"name":"Ensuite Glass","UUID":"1803449b-0cc0-4cd7-8139-6e853b07d5e3","serviceName":"Switch","config":{"id":"6a013907.a43d28","type":"homekit-service","z":"e6d22196.bc1ac","isParent":true,"bridge":"8cb8961e.cbe378","parentService":"","name":"Ensuite Glass","serviceName":"Switch","topic":"","filter":false,"manufacturer":"NRCHKB","model":"0.130.29","serialNo":"Default Serial Number","firmwareRev":"0.130.29","hardwareRev":"0.130.29","softwareRev":"0.130.29","cameraConfigVideoProcessor":"ffmpeg","cameraConfigSource":"","cameraConfigStillImageSource":"","cameraConfigMaxStreams":2,"cameraConfigMaxWidth":1280,"cameraConfigMaxHeight":720,"cameraConfigMaxFPS":10,"cameraConfigMaxBitrate":300,"cameraConfigVideoCodec":"libx264","cameraConfigAudioCodec":"libfdk_aac","cameraConfigAudio":false,"cameraConfigPacketSize":1316,"cameraConfigVerticalFlip":false,"cameraConfigHorizontalFlip":false,"cameraConfigMapVideo":"0:0","cameraConfigMapAudio":"0:1","cameraConfigVideoFilter":"scale=1280:720","cameraConfigAdditionalCommandLine":"-tune zerolatency","cameraConfigDebug":false,"cameraConfigSnapshotOutput":"disabled","cameraConfigInterfaceName":"","characteristicProperties":"{}","waitForSetupMsg":false,"outputs":2,"x":370,"y":480,"wires":[["ef0f1e5e.0cc51"],[]]}} +19ms
  NRCHKB:ServiceUtils Looking for service with UUID '1803449b-0cc0-4cd7-8139-6e853b07d5e3'... +0ms
  NRCHKB:ServiceUtils ... didn't find it. Adding new Switch service. +0ms
  NRCHKB:HAPBridgeNode Publishing bridge with name 'Homered Bridge', pin code '123-44-925' and 2 accessories. +7s
  HAP-NodeJS:Accessory [Homered Bridge] Creating new IdentifierCache +5s
```

EXCEPT, the very next line is now:

```
9 Feb 11:49:30 - [error] [homekit-service:Ensuite Glass] Bridge [8cb8961e.cbe378] publish failed due to: TypeError: characteristic._assignID is not a function
```

**Comment by macinspak** [comment]
OK, that's definitely it. I don't know how this hasn't been a major issue before.

So essentially I had to go through the code and change a LOT of:
```javascript
for (a in b) {
}
```

Array iterators to the better:
```javascript
b.forEach (function(x) {
});
```

And now I can add the bridge again. So maybe it was just one library that adds Array.prototype.isEqual = function() ...
but that just breaks it all unless HAP-nodejs is smarter and uses a more robust array iterator throughout the entire library.

I've forked the repo to do the patch, but I might not get to it immediately, I'll see how I go.

**Comment by macinspak** [comment]
OK, TS throwing a lot of errors with my mods, and not sure how to compile to test. Can someone take this on? I have the mods I made in https://github.com/Anamico/HAP-NodeJS. Just need someone with a little more knowledge of this project to clean up the build errors.

**Comment by macinspak** [comment]
Thanks for the offer to help @Shaquu I've added you as a maintainer.
I committed the changes to master before thinking, sorry, so we may need to just branch it and re-apply, then get it tested and working and submit the pull request.

I made a bunch of changes throughout to places I thought the issue could occur, pretty much any object key iterator was changed. It simplified code due to the iterator forEach pulling the value not the index, but it may not be working fine with npm test as it may be not detecting the object class properly.

Also, in some cases where it is looping through to find the match, then deleting it (for example), I switched them to .find logic with the same conditions in the filter.

I believe they should have the same intended post conditions for the searches and the return values on no match, but we just need to check each I guess for sanity and then make sure it's compiling. If you can let me know how to build it to insert into my node-red instance, then I can test heavily.

The basic changes I made to prove the fix in my instance work to pair to homekit and take instructions, but the final state is stuck (not updating), but the changes in the repo included other sections, so once I test it I can see if it works for my use cases at least.

Thanks again for the help.

**Comment by otaku** [comment]
@macinspak I went the easier route and switched everything from for...in to for...of

I haven't actually tested it yet due to lack of time, but it is compiling and passing all jest tests without any typescript errors.

https://github.com/otaku/HAP-NodeJS

**Comment by macinspak** [comment]
Great idea, much easier.

I made the same changes to my prod environment (house automation) and it's working perfectly! Maybe it's actually the same issue with why my homebridge appliances are all reporting "no response". Might look into that later.

So I am confident it is the right fix, but cannot assist in testing. So can someone please pick this up and get it into the next release as I expect it's impacting a LOT more people than we think at this stage as it is a fundamental flaw that is easily fixed.

**Comment by Shaquu** [comment]
@macinspak I can see you removed your fork?
@otaku can you rebase your changes on beta branch? I can see you are missing some new code parts.

**Comment by macinspak** [comment]
Yes, @otaku had done the fix a simpler way and so it was duplicate work.

**Comment by macinspak** [comment]
Sorry, may have been @shaquu

**Comment by otaku** [comment]
Changes were applied to beta branch.  I went ahead and did a few more non-breaking refactors.  PR opened #879

**Comment by Supereg** [comment]
Is this issue solved, as #879 is now merged into the latest beta build?

**Comment by otaku** [comment]
@macinspak can verify, but I believe so.  I was getting errors after installing a lot of node-red modules.  I restored to a backup so I no longer can reproduce.

**Comment by macinspak** [comment]
Sorry, how do I deploy the beta to my instance? Can I use nom directly? Or do I need to run a script to build it? I had trouble before as it would not deploy the build folder

**Comment by Supereg** [comment]
Actually the fix is now live with the just now released 0.9.0 version. Not sure if that helps you with regards to your deployment problem. I didn't really closely follow this issue, so I don't really have an idea about your setup 😅

**Comment by macinspak** [comment]
Cool, will endeavor to test in next few days and report back. Thanks.
